### ISSUE TITLE
Duplicated clang diagnostics with same message

### ISSUE THREAD

**Issue by strimo378** [llvm/llvm-project#83955]
Hi all,

I am currently analyzing clang diagnostics and found several diagnostics that have exactly the same message. For warnings and downgradable errors, I have not check if they have the same -W flag that is displayed alongside with the diagnostics. I have also not check if there are warnings/errors with the same message (since warning can be turned into errors).

Warning: Duplicate message: `error:  TLS model '%0' is not yet supported on AIX`
id1: err_aix_unsupported_tls_model
id2: err_aix_attr_unsupported_tls_model

Warning: Duplicate message: `error:  for range declaration must declare a variable`
id1: err_for_range_expected_decl
id2: err_for_range_decl_must_be_var

Warning: Duplicate message: `error:  redefinition of module '%0'`
id1: err_module_redefinition
id2: err_mmap_module_redefinition

Warning: Duplicate message: `error:  ambiguous conversion from type %0 to an integral or unscoped enumeration type`
id1: err_omp_ambiguous_conversion
id2: err_ice_ambiguous_conversion

Warning: Duplicate message: `error:  expected 'begin' or 'end'`
id1: err_pp_assume_nonnull_syntax
id2: err_pp_arc_cf_code_audited_syntax

Warning: Duplicate message: `error:  friend declaration of %0 does not match any declaration in %1`
id1: err_qualified_friend_no_match
id2: err_friend_decl_does_not_match

Warning: Duplicate message: `error:  unknown target triple '%0'`
id1: err_target_unknown_triple
id2: err_drv_unknown_target_triple

Warning: Duplicate message: `error:  %0 is not supported on this target`
id1: err_type_unsupported
id2: err_keyword_not_supported_on_target

Warning: Duplicate message: `error:  expression is not assignable`
id1: err_typecheck_expression_not_modifiable_lvalue
id2: err_no_subobject_property_setting

Warning: Duplicate message: `error:  cannot find interface declaration for %0`
id1: err_undef_interface
id2: err_no_nsconstant_string_class

Warning: Duplicate message: `error:  in-class initializer for static data member of type %0 requires 'constexpr' specifier`
id1: ext_in_class_initializer_float_type_cxx11
id2: err_in_class_initializer_literal_type

Warning: Duplicate message: `warning:  static %select{function|variable}0 %1 is used in an inline function with external linkage`
id1: ext_internal_in_extern_inline_quiet
id2: ext_internal_in_extern_inline

Warning: Duplicate message: `error:  pasting formed '%0', an invalid preprocessing token`
id1: ext_pp_bad_paste_ms
id2: err_pp_bad_paste

Warning: Duplicate message: `error:  taking the address of a temporary object of type %0`
id1: ext_typecheck_addrof_temporary
id2: err_typecheck_addrof_temporary

Warning: Duplicate message: `error:  incompatible function pointer types %select{%diff{assigning to $ from $|assigning to different types}0,1|%diff{passing $ to parameter of type $|passing to parameter of different type}0,1|%diff{returning $ from a function with result type $|returning from function with different return type}0,1|%diff{converting $ to type $|converting between types}0,1|%diff{initializing $ with an expression of type $|initializing with expression of different type}0,1|%diff{sending $ to parameter of type $|sending to parameter of different type}0,1|%diff{casting $ to type $|casting between types}0,1}2%select{|; dereference with *|; take the address with &|; remove *|; remove &}3`
id1: ext_typecheck_convert_incompatible_function_pointer
id2: err_typecheck_convert_incompatible_function_pointer

Warning: Duplicate message: `error:  incompatible integer to pointer conversion %select{%diff{assigning to $ from $|assigning to different types}0,1|%diff{passing $ to parameter of type $|passing to parameter of different type}0,1|%diff{returning $ from a function with result type $|returning from function with different return type}0,1|%diff{converting $ to type $|converting between types}0,1|%diff{initializing $ with an expression of type $|initializing with expression of different type}0,1|%diff{sending $ to parameter of type $|sending to parameter of different type}0,1|%diff{casting $ to type $|casting between types}0,1}2%select{|; dereference with *|; take the address with &|; remove *|; remove &}3`
id1: ext_typecheck_convert_int_pointer
id2: err_typecheck_convert_int_pointer

Warning: Duplicate message: `error:  incompatible pointer to integer conversion %select{%diff{assigning to $ from $|assigning to different types}0,1|%diff{passing $ to parameter of type $|passing to parameter of different type}0,1|%diff{returning $ from a function with result type $|returning from function with different return type}0,1|%diff{converting $ to type $|converting between types}0,1|%diff{initializing $ with an expression of type $|initializing with expression of different type}0,1|%diff{sending $ to parameter of type $|sending to parameter of different type}0,1|%diff{casting $ to type $|casting between types}0,1}2%select{|; dereference with *|; take the address with &|; remove *|; remove &}3`
id1: ext_typecheck_convert_pointer_int
id2: err_typecheck_convert_pointer_int

Warning: Duplicate message: `warning:  duplicate '%0' declaration specifier`
id1: ext_warn_duplicate_declspec
id2: ext_duplicate_declspec

Warning: Duplicate message: `error:  '%0' cannot be signed or unsigned`
id1: ext_wchar_t_sign_spec
id2: err_invalid_sign_spec

Warning: Duplicate message: `note:  %select{and|because}0 '%1' evaluated to false`
id1: note_concept_specialization_constraint_evaluated_to_false
id2: note_atomic_constraint_evaluated_to_false

Warning: Duplicate message: `note:  did you mean '%0'?`
id1: note_doc_tparam_name_suggestion
id2: note_doc_param_name_suggestion

Warning: Duplicate message: `note:  previous documentation`
id1: note_doc_tparam_previous
id2: note_doc_param_previous

Warning: Duplicate message: `note:  %0 declared here`
id1: note_entity_declared_at
id2: note_callee_decl

Warning: Duplicate message: `note:  %select{and|because}0 '%1' would be invalid: %2`
id1: note_expr_requirement_type_requirement_substitution_error
id2: note_expr_requirement_expr_substitution_error

Warning: Duplicate message: `note:  %select{and|because}0 '%1' would be invalid`
id1: note_expr_requirement_type_requirement_unknown_substitution_error
id2: note_expr_requirement_expr_unknown_substitution_error

Warning: Duplicate message: `note:  conversion to %select{integral|enumeration}0 type %1 declared here`
id1: note_ice_conversion_here
id2: note_array_size_conversion

Warning: Duplicate message: `note:  %0 declared here`
id1: note_musttail_callee_defined_here
id2: note_entity_declared_at

Warning: Duplicate message: `note:  %select{and|because}0 '%1' would be invalid`
id1: note_nested_requirement_unknown_substitution_error
id2: note_expr_requirement_type_requirement_unknown_substitution_error

Warning: Duplicate message: `note:  %0 declared here`
id1: note_objc_circular_container_declared_here
id2: note_musttail_callee_defined_here

Warning: Duplicate message: `note:  conversion to %select{integral|enumeration}0 type %1 declared here`
id1: note_omp_conversion_here
id2: note_ice_conversion_here

Warning: Duplicate message: `note:  previously defined here`
id1: note_prev_module_definition
id2: note_mmap_prev_definition

Warning: Duplicate message: `note:  %0 declared here`
id1: note_previous_decl
id2: note_objc_circular_container_declared_here

Warning: Duplicate message: `note:  declared here`
id1: note_shadow_field
id2: note_declared_at

Warning: Duplicate message: `note:  %select{and|because}0 '%1' would be invalid: %2`
id1: note_type_requirement_substitution_error
id2: note_expr_requirement_type_requirement_substitution_error

Warning: Duplicate message: `note:  %select{and|because}0 '%1' would be invalid`
id1: note_type_requirement_unknown_substitution_error
id2: note_nested_requirement_unknown_substitution_error

Warning: Duplicate message: `warning:  non-portable path to file '%0'; specified path differs in case from file name on disk`
id1: pp_nonportable_system_path
id2: pp_nonportable_path

Warning: Duplicate message: `error:  argument value %0 is outside the valid range [%1, %2]`
id1: warn_argument_invalid_range
id2: err_argument_invalid_range

Warning: Duplicate message: `warning:  %select{arithmetic between|bitwise operation between|comparison of|conditional expression between|compound assignment of}0 different enumeration types%diff{ ($ and $)|}1,2`
id1: warn_arith_conv_mixed_enum_types
id2: warn_arith_conv_mixed_anon_enum_types

Warning: Duplicate message: `warning:  %select{arithmetic between|bitwise operation between|comparison of|conditional expression between|compound assignment of}0 different enumeration types%diff{ ($ and $)|}1,2 is deprecated`
id1: warn_arith_conv_mixed_enum_types_cxx20
id2: warn_arith_conv_mixed_anon_enum_types_cxx20

Warning: Duplicate message: `warning:  cast %diff{from $ to $ |}0,1converts to incompatible function type`
id1: warn_cast_function_type_strict
id2: warn_cast_function_type

Warning: Duplicate message: `warning:  %select{arithmetic between|bitwise operation between|comparison of|conditional expression between|compound assignment of}0 different enumeration types%diff{ ($ and $)|}1,2`
id1: warn_comparison_mixed_enum_types
id2: warn_arith_conv_mixed_enum_types

Warning: Duplicate message: `warning:  %select{arithmetic between|bitwise operation between|comparison of|conditional expression between|compound assignment of}0 different enumeration types%diff{ ($ and $)|}1,2 is deprecated`
id1: warn_comparison_mixed_enum_types_cxx20
id2: warn_arith_conv_mixed_enum_types_cxx20

Warning: Duplicate message: `warning:  using the result of an assignment as a condition without parentheses`
id1: warn_condition_is_idiomatic_assignment
id2: warn_condition_is_assignment

Warning: Duplicate message: `warning:  %select{arithmetic between|bitwise operation between|comparison of|conditional expression between|compound assignment of}0 different enumeration types%diff{ ($ and $)|}1,2`
id1: warn_conditional_mixed_enum_types
id2: warn_comparison_mixed_enum_types

Warning: Duplicate message: `warning:  %select{arithmetic between|bitwise operation between|comparison of|conditional expression between|compound assignment of}0 different enumeration types%diff{ ($ and $)|}1,2 is deprecated`
id1: warn_conditional_mixed_enum_types_cxx20
id2: warn_comparison_mixed_enum_types_cxx20

Warning: Duplicate message: `warning:  conflicting variadic declaration of method and its implementation`
id1: warn_conflicting_variadic
id2: warn_conflicting_overriding_variadic

Warning: Duplicate message: `warning:  declaration shadows a %select{local variable|variable in %2|static data member of %2|field of %2|typedef in %2|type alias in %2|structured binding}1`
id1: warn_decl_shadow_uncaptured_local
id2: warn_decl_shadow

Warning: Duplicate message: `warning:  macro expansion producing 'defined' has undefined behavior`
id1: warn_defined_in_object_type_macro
id2: warn_defined_in_function_type_macro

Warning: Duplicate message: `warning:  argument '%0' requires profile-guided optimization information`
id1: warn_drv_diagnostics_misexpect_requires_pgo
id2: warn_drv_diagnostics_hotness_requires_pgo

Warning: Duplicate message: `warning:  duplicate '%0' declaration specifier`
id1: warn_duplicate_declspec
id2: ext_warn_duplicate_declspec

Warning: Duplicate message: `warning:  %select{values of type|enum values with underlying type}2 '%0' should not be used as format arguments; add an explicit cast to %1 instead`
id1: warn_format_argument_needs_cast_pedantic
id2: warn_format_argument_needs_cast

Warning: Duplicate message: `warning:  format specifies type %0 but the argument has %select{type|underlying type}2 %1`
id1: warn_format_conversion_argument_type_mismatch_confusion
id2: warn_format_conversion_argument_type_mismatch

Warning: Duplicate message: `warning:  format specifies type %0 but the argument has %select{type|underlying type}2 %1`
id1: warn_format_conversion_argument_type_mismatch_pedantic
id2: warn_format_conversion_argument_type_mismatch_confusion

Warning: Duplicate message: `warning:  '%0' will always overflow; destination buffer has size %1, but size argument is %2`
id1: warn_fortify_source_overflow
id2: warn_builtin_chk_overflow

Warning: Duplicate message: `warning:  calling function %1 requires holding %0 %select{'%2'|'%2' exclusively}3`
id1: warn_fun_requires_lock_precise
id2: warn_fun_requires_lock

Warning: Duplicate message: `warning:  implicit conversion loses integer precision: %0 to %1`
id1: warn_impcast_integer_precision
id2: warn_impcast_integer_64_32

Warning: Duplicate message: `warning:  implicit conversion from %2 to %3 changes value from %0 to %1`
id1: warn_impcast_integer_precision_constant
id2: warn_impcast_integer_float_precision_constant

Warning: Duplicate message: `warning:  implicit conversion from %0 to %1 changes value from %2 to %3`
id1: warn_impcast_literal_float_to_integer
id2: warn_impcast_float_to_integer

Warning: Duplicate message: `warning:  implicit conversion of out of range value from %0 to %1 is undefined`
id1: warn_impcast_literal_float_to_integer_out_of_range
id2: warn_impcast_float_to_integer_out_of_range

Warning: Duplicate message: `warning:  initializer %select{partially |}0overrides prior initialization of this subobject`
id1: warn_initializer_overrides
id2: ext_initializer_overrides

Warning: Duplicate message: `warning:  cast to %1 from smaller integer type %0`
id1: warn_int_to_void_pointer_cast
id2: warn_int_to_pointer_cast

Warning: Duplicate message: `warning:  overflow in expression; result is %0 with type %1`
id1: warn_integer_constant_overflow
id2: warn_fixedpoint_constant_overflow

Warning: Duplicate message: `warning:  multi-character character constant`
id1: warn_multichar_character_literal
id2: warn_four_char_character_literal

Warning: Duplicate message: `warning:  extra qualification on member %0`
id1: warn_namespace_member_extra_qualification
id2: warn_member_extra_qualification

Warning: Duplicate message: `warning:  no newline at end of file`
id1: warn_no_newline_eof
id2: ext_no_newline_eof

Warning: Duplicate message: `warning:  bitmasking for introspection of Objective-C object pointers is strongly discouraged`
id1: warn_objc_pointer_masking_performSelector
id2: warn_objc_pointer_masking

Warning: Duplicate message: `warning:  cast to smaller integer type %1 from %0`
id1: warn_pointer_to_int_cast
id2: warn_pointer_to_enum_cast

Warning: Duplicate message: `warning:  %0 is not defined, evaluates to 0`
id1: warn_pp_undef_prefix
id2: warn_pp_undef_identifier

Warning: Duplicate message: `warning:  operator '?:' has lower precedence than '%0'; '%0' will be evaluated first`
id1: warn_precedence_conditional
id2: warn_precedence_bitwise_conditional

Warning: Duplicate message: `error:  non-void %select{function|method}1 %0 should return a value`
id1: warn_return_missing_expr
id2: ext_return_missing_expr

Warning: Duplicate message: `warning:  explicitly assigning value of variable of type %0 to itself%select{|; did you mean to assign to member %2?}1`
id1: warn_self_assignment_overloaded
id2: warn_self_assignment_builtin

Warning: Duplicate message: `warning:  multiple methods named %0 found`
id1: warn_strict_multiple_method_decl
id2: warn_multiple_method_decl

Warning: Duplicate message: `warning:  @selector expression formed with potentially direct selector %0`
id1: warn_strict_potentially_direct_selector_expression
id2: warn_potentially_direct_selector_expression

Warning: Duplicate message: `warning:  %0 overrides a destructor but is not marked 'override'`
id1: warn_suggest_destructor_marked_not_override_overriding
id2: warn_inconsistent_destructor_marked_not_override_overriding

Warning: Duplicate message: `warning:  %0 overrides a member function but is not marked 'override'`
id1: warn_suggest_function_marked_not_override_overriding
id2: warn_inconsistent_function_marked_not_override_overriding

Warning: Duplicate message: `warning:  result of comparison of %select{constant %0|true|false}1 with %select{expression of type %2|boolean expression}3 is always %4`
id1: warn_tautological_bool_compare
id2: warn_out_of_range_compare

Warning: Duplicate message: `warning:  ordered comparison of function pointers (%0 and %1)`
id1: warn_typecheck_ordered_comparison_of_function_pointers
id2: ext_typecheck_ordered_comparison_of_function_pointers

Warning: Duplicate message: `warning:  %0 is only available on %1 %2 or newer`
id1: warn_unguarded_availability_new
id2: warn_unguarded_availability

Warning: Duplicate message: `warning:  ignoring return value of function declared with %0 attribute`
id1: warn_unused_result
id2: warn_unused_call

Warning: Duplicate message: `warning:  unused variable %0`
id1: warn_unused_variable
id2: warn_unused_const_variable

Warning: Duplicate message: `warning:  %select{reading|writing}3 the value pointed to by %1 requires holding %0 %select{'%2'|'%2' exclusively}3`
id1: warn_var_deref_requires_lock_precise
id2: warn_var_deref_requires_lock

Warning: Duplicate message: `warning:  %select{reading|writing}3 variable %1 requires holding %0 %select{'%2'|'%2' exclusively}3`
id1: warn_variable_requires_lock_precise
id2: warn_variable_requires_lock

Warning: Duplicate message: `warning:  cast to smaller integer type %1 from %0`
id1: warn_void_pointer_to_enum_cast
id2: warn_pointer_to_int_cast

Warning: Duplicate message: `warning:  cast to smaller integer type %1 from %0`
id1: warn_void_pointer_to_int_cast
id2: warn_void_pointer_to_enum_cast

Warning: Duplicate message: `warning:  %select{|empty }0%select{struct|union}1 has size 0 in C, %select{size 1|non-zero size}2 in C++`
id1: warn_zero_size_struct_union_in_extern_c
id2: warn_zero_size_struct_union_compat

**Comment by AaronBallman** [comment]
FWIW, this is not unexpected (though I would not be surprised if there were cases where we should combine diagnostics). For example:
```
Warning: Duplicate message: error:  for range declaration must declare a variable
id1: err_for_range_expected_decl
id2: err_for_range_decl_must_be_var
```
`err_for_range_expected_decl` is a parsing error (`DiagnosticParseKinds.td`) while `err_for_range_decl_must_be_var` is a semantic error (`DiagnosticSemaKinds.td`), so the diagnostics are issued from different components in the compiler. However, we could add one of them to `DiagnosticCommonKinds.td` and use the same diagnostic in both places. But there are likely other diagnostics where they should remain separate (for example, warnings under different diagnostic groups for <reasons>).

**Comment by strimo378** [comment]
@AaronBallman thank you for your response.

I think the problem of duplicated diagnostics raise also from a documentation point of view. End users can only identify a diagnostics by their text as e.g. shown in https://clang.llvm.org/docs/DiagnosticsReference.html . For instance, the two warnings ext_no_newline_eof and warn_no_newline_eof have the same text but are in different groups. ext_no_newline_eof is part of -Wpedantic while warn_no_newline_eof is not. warn_no_newline_eof is used for C++ and ext_no_newline_eof is used for C.
